internal/plugin/plugin.go

```diff
@@ -0,0 +1,257 @@
+	uploadPath := filePath + ".incomplete"
+
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, fmt.Errorf("file already exists: %s", sanitizedFilename)
+	}
+
+	var alreadyUploadedBytes int64 = 0
+	if stat, err := os.Stat(uploadPath); err == nil {
+		alreadyUploadedBytes = stat.Size()
+	}
+
+	uploadId := "plugin_" + uuid.New().String()
+	file, err := os.OpenFile(uploadPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file for upload: %v", err)
+	}
+
+	storage.AddPendingUpload(uploadId, storage.PendingUpload{
+		File:                 file,
+		Size:                 size,
+		AlreadyUploadedBytes: alreadyUploadedBytes,
+	})
+
+	return &storage.StorageFileUpload{
+		AlreadyUploadedBytes: alreadyUploadedBytes,
+		DataChannel:          uploadId,
+	}, nil
+}
+
+func RpcPluginExtract(filename string) (*PluginManifest, error) {
+	sanitizedFilename, err := storage.SanitizeFilename(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	filePath := path.Join(pluginsUploadFolder, sanitizedFilename)
+	extractFolder, err := extractPlugin(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.Remove(filePath); err != nil {
+		return nil, fmt.Errorf("failed to delete uploaded file: %v", err)
+	}
+
+	manifest, err := readManifest(extractFolder)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get existing PluginInstall
+	install, ok := pluginDatabase.Plugins[manifest.Name]
+	if !ok {
+		install = &PluginInstall{
+			Enabled:           false,
+			Version:           manifest.Version,
+			ExtractedVersions: make(map[string]string),
+		}
+	}
+
+	_, ok = install.ExtractedVersions[manifest.Version]
+	if ok {
+		return nil, fmt.Errorf("this version has already been uploaded: %s", manifest.Version)
+	}
+
+	install.ExtractedVersions[manifest.Version] = extractFolder
+	pluginDatabase.Plugins[manifest.Name] = install
+
+	if err := pluginDatabase.Save(); err != nil {
+		return nil, fmt.Errorf("failed to save plugin database: %v", err)
+	}
+
+	return manifest, nil
+}
+
+func RpcPluginInstall(name string, version string) error {
+	pluginInstall, ok := pluginDatabase.Plugins[name]
+	if !ok {
+		return fmt.Errorf("plugin not found: %s", name)
+	}
+
+	if pluginInstall.Version == version && pluginInstall.Enabled {
+		fmt.Printf("Plugin %s is already installed with version %s\n", name, version)
+		return nil
+	}
+
+	_, ok = pluginInstall.ExtractedVersions[version]
+	if !ok {
+		return fmt.Errorf("plugin version not found: %s", version)
+	}
+
+	pluginInstall.Version = version
+	pluginInstall.Enabled = true
+	pluginDatabase.Plugins[name] = pluginInstall
+
+	if err := pluginDatabase.Save(); err != nil {
+		return fmt.Errorf("failed to save plugin database: %v", err)
+	}
+
+	err := pluginInstall.ReconcileSubprocess()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin %s: %v", name, err)
+	}
+
+	// TODO: Determine if the old extract should be removed
```

Comment:
Maybe simplify life by just offering a "cleanup temporaries" button that deletes all historical/inactive versions and also [any partial uploads](https://github.com/jetkvm/kvm/pull/10/files#diff-034908473a1bf69a8b56962feef4c44ae39cf2ae61eebe107cf0a05d19a6809fR47) and [temp files](https://github.com/jetkvm/kvm/pull/10/files#diff-c8b8c056089c83b8be8f143dec61eb13aef9f1bd3c9d38c7b685566fb372e577R44)